synapse/handlers/federation.py

```diff
@@ -1667,16 +1667,52 @@ async def on_send_join_request(self, origin: str, pdu: EventBase) -> JsonDict:
         # would introduce the danger of backwards-compatibility problems.
         event.internal_metadata.send_on_behalf_of = origin
 
-        context = await self._handle_new_event(origin, event)
+        # Calculate the event context.
+        context = await self._prep_event(
+            origin, event, state=None, auth_events=None, backfilled=False
+        )
```

Comment:
hrhrm. `_prep_event` has some side-effects that I'm not *sure* we want to do just here. we might want to split it in half (and see if we can think of better names while we're at it.)

Reply:
From my reading, the parts we might not want to do are the call to `maybe_kick_guest_users` and maybe caching joined hosts? Does that match what you were thinking?

Reply:
I think I would split it after `do_auth` (ie, at https://github.com/matrix-org/synapse/blob/94bdb01a90f09a67b2d83d1af43be060676f354c/synapse/handlers/federation.py#L2269). `do_auth` is only called there, so I think you can move the first half of `_prep_event` into `do_auth`.